File: dev/ci/scripts/utils/Copilot/mcp_server_python/mcp-server.py
#!/usr/bin/env python3
"""
Model Context Protocol (MCP) Server for Global Workflow
Provides context about the global-workflow repository structure and functionality
"""

import json
import sys
import asyncio
from typing import Any, Dict, List, Optional
import os
import logging
from pathlib import Path

# Set up logging for debugging (writes to stderr so it doesn't interfere with MCP protocol)
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    stream=sys.stderr
)

class GlobalWorkflowMCPServer:
    """MCP Server for Global Workflow repository"""
    
    def __init__(self):
        """Initialize the MCP server"""
        self.repo_root = Path(__file__).parent
        self.logger = logging.getLogger("GlobalWorkflowMCP")
        self.logger.info(f"Initializing MCP server for repository: {self.repo_root}")
        
        self.tools = [
            {
                "name": "get_workflow_info",
                "description": "Get information about the global workflow system",
                "inputSchema": {
                    "type": "object",
                    "properties": {
                        "component": {
                            "type": "string",
                            "description": "Component to get info about (jobs, scripts, configs, etc.)"
                        }
                    }
                }
            },
            {
                "name": "list_job_scripts",
                "description": "List available job scripts in the workflow",
                "inputSchema": {
                    "type": "object",
                    "properties": {}
                }
            },
            {
                "name": "get_config_info",
                "description": "Get configuration information for different systems",
                "inputSchema": {
                    "type": "object",
                    "properties": {
                        "system": {
                            "type": "string",
                            "description": "System name (hera, orion, wcoss2, etc.)"
                        }
                    }
                }
            }
        ]
    
    async def handle_request(self, request: Dict[str, Any]) -> Dict[str, Any]:
        """Handle incoming MCP requests"""
        method = request.get("method")
        params = request.get("params", {})
        
        self.logger.info(f"Handling request: {method}")
        
        if method == "initialize":
            self.logger.info("Server initialized successfully")
            return {
                "protocolVersion": "2024-11-05",
                "capabilities": {
                    "tools": {}
                },
                "serverInfo": {
                    "name": "global-workflow-mcp",
                    "version": "1.0.0"
                }
            }
        
        elif method == "tools/list":
            self.logger.info(f"Listing {len(self.tools)} available tools")
            return {"tools": self.tools}
        
        elif method == "tools/call":
            tool_name = params.get("name")
            arguments = params.get("arguments", {})
            
            self.logger.info(f"Calling tool: {tool_name} with args: {arguments}")
            
            if tool_name == "get_workflow_info":
                return await self._get_workflow_info(arguments)
            elif tool_name == "list_job_scripts":
                return await self._list_job_scripts(arguments)
            elif tool_name == "get_config_info":
                return await self._get_config_info(arguments)
        
        self.logger.warning(f"Unknown method: {method}")
        return {"error": f"Unknown method: {method}"}
    
    async def _get_workflow_info(self, args: Dict[str, Any]) -> Dict[str, Any]:
        """Get workflow information"""
        component = args.get("component")
        
        info = {
            "overview": "NOAA Global Workflow - Operational weather prediction system",
            "description": "The Global Workflow is NOAA's end-to-end numerical weather prediction system that runs operationally to produce GFS, GDAS, and GEFS forecasts.",
            "workflow_structure": {
                "data_assimilation": "GDAS (Global Data Assimilation System) - combines observations with model background",
                "deterministic_forecast": "GFS (Global Forecast System) - single deterministic forecast",
                "ensemble_forecast": "GEFS (Global Ensemble Forecast System) - probabilistic ensemble forecasts",
                "workflow_engine": "Rocoto XML-based workflow management with job dependencies"
            },
            "components": {
                "jobs": "Job scripts (J*) for various workflow tasks - batch job submission scripts",
                "scripts": "Execution scripts (ex*) for workflow execution - actual implementation logic",
                "ush": "Utility shell scripts and common functions",
                "parm": "Parameter files, configurations, and templates",
                "fix": "Fixed input data files, lookup tables, and static datasets",
                "sorc": "Source code directories for models and utilities",
                "modulefiles": "Environment module files for different HPC systems",
                "env": "System-specific environment configuration files"
            },
            "systems": ["hera", "orion", "hercules", "wcoss2", "gaeac5", "gaeac6"],
            "runs": {
                "gfs": "Global Forecast System - deterministic weather forecasts",
                "gdas": "Global Data Assimilation System - analysis and reanalysis",
                "gefs": "Global Ensemble Forecast System - ensemble probabilistic forecasts"
            },
            "rocoto": "Workflow management using Rocoto XML with job dependencies and scheduling",
            "directory_structure": {
                "/jobs": "Production job scripts (JGDAS_*, JGFS_*, JGLOBAL_*)",
                "/scripts": "Execution scripts (exgdas_*, exgfs_*, exglobal_*)",
                "/dev/jobs": "Development job scripts",
                "/ush": "Utility scripts and functions",
                "/parm": "Configuration files and parameters",
                "/fix": "Static input data and lookup tables",
                "/sorc": "Source code for UFS, GSI, and other components"
            }
        }
        
        if component and component in info:
            return {"content": [{"type": "text", "text": str(info[component])}]}
        else:
            return {"content": [{"type": "text", "text": json.dumps(info, indent=2)}]}
    
    async def _list_job_scripts(self, args: Dict[str, Any]) -> Dict[str, Any]:
        """List available job scripts"""
        jobs_dir = self.repo_root / "jobs"
        if not jobs_dir.exists():
            return {"content": [{"type": "text", "text": "Jobs directory not found"}]}
        
        job_scripts = []
        for job_file in jobs_dir.glob("J*"):
            if job_file.is_file():
                job_scripts.append(job_file.name)
        
        job_list = "\n".join(sorted(job_scripts))
        return {"content": [{"type": "text", "text": f"Available job scripts:\n{job_list}"}]}
    
    async def _get_config_info(self, args: Dict[str, Any]) -> Dict[str, Any]:
        """Get configuration information"""
        system = args.get("system", "general")
        
        config_info = {
            "hera": "NOAA RDHPCS Hera system configuration",
            "orion": "NOAA RDHPCS Orion system configuration", 
            "hercules": "NOAA RDHPCS Hercules system configuration",
            "wcoss2": "NOAA WCOSS2 operational system configuration",
            "general": "Configuration files are in parm/ directory"
        }
        
        info = config_info.get(system, f"Configuration for {system}")
        return {"content": [{"type": "text", "text": info}]}

async def main():
    """Main MCP server loop"""
    server = GlobalWorkflowMCPServer()
    
    # Read from stdin and write to stdout for MCP protocol
    while True:
        try:
            line = await asyncio.get_event_loop().run_in_executor(None, sys.stdin.readline)
            if not line:
                break
            
            request = json.loads(line.strip())
            response = await server.handle_request(request)
            
            # Add request ID if present
            if "id" in request:
                response["id"] = request["id"]
            
            print(json.dumps(response))
            sys.stdout.flush()
            
        except json.JSONDecodeError:
            continue
        except Exception as e:
            error_response = {
                "error": {
                    "code": -32603,
                    "message": str(e)
                }
            }
            if "id" in locals() and "request" in locals() and "id" in request:
                error_response["id"] = request["id"]
            
            print(json.dumps(error_response))
            sys.stdout.flush()

if __name__ == "__main__":
    asyncio.run(main())

File: mcp-server.js
#!/usr/bin/env node

/**
 * Model Context Protocol (MCP) Server for Global Workflow
 * Provides context about the global-workflow repository structure and functionality
 */

import { Server } from '@modelcontextprotocol/sdk/server/index.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
} from '@modelcontextprotocol/sdk/types.js';
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * Create and configure the MCP server
 */
function createServer() {
  const server = new Server(
    {
      name: 'global-workflow-mcp',
      version: '1.0.0',
    },
    {
      capabilities: {
        tools: {},
      },
    }
  );

  /**
   * List available tools
   */
  server.setRequestHandler(ListToolsRequestSchema, async () => {
    return {
      tools: [
        {
          name: 'get_workflow_structure',
          description: 'Get the structure and overview of the global workflow system',
          inputSchema: {
            type: 'object',
            properties: {
              component: {
                type: 'string',
                description: 'Specific component to focus on (optional)',
                enum: ['jobs', 'scripts', 'configs', 'overview']
              }
            }
          }
        },
        {
          name: 'list_job_scripts',
          description: 'List all available job scripts in the workflow',
          inputSchema: {
            type: 'object',
            properties: {}
          }
        },
        {
          name: 'get_system_configs',
          description: 'Get configuration information for different HPC systems',
          inputSchema: {
            type: 'object',
            properties: {
              system: {
                type: 'string',
                description: 'HPC system name',
                enum: ['hera', 'orion', 'hercules', 'wcoss2', 'gaeac5', 'gaeac6']
              }
            }
          }
        },
        {
          name: 'explain_workflow_component',
          description: 'Explain a specific workflow component or directory',
          inputSchema: {
            type: 'object',
            properties: {
              component: {
                type: 'string',
                description: 'Component name (e.g., rocoto, gsi, ufs)',
                required: true
              }
            },
            required: ['component']
          }
        }
      ]
    };
  });

  /**
   * Handle tool calls
   */
  server.setRequestHandler(CallToolRequestSchema, async (request) => {
    const { name, arguments: args } = request.params;

    try {
      switch (name) {
        case 'get_workflow_structure':
          return await getWorkflowStructure(args?.component);
        
        case 'list_job_scripts':
          return await listJobScripts();
        
        case 'get_system_configs':
          return await getSystemConfigs(args?.system);
        
        case 'explain_workflow_component':
          return await explainWorkflowComponent(args?.component);
        
        default:
          throw new Error(`Unknown tool: ${name}`);
      }
    } catch (error) {
      return {
        content: [
          {
            type: 'text',
            text: `Error: ${error.message}`
          }
        ]
      };
    }
  });

  return server;
}

/**
 * Get workflow structure information
 */
async function getWorkflowStructure(component) {
  const structure = {
    overview: `
Global Workflow - NOAA's Operational Weather Prediction System

Key Components:
- Jobs: Batch job scripts for various workflow tasks
- Scripts: Shell scripts that implement job functionality  
- USH: Utility shell scripts and common functions
- Parm: Parameter files and configuration templates
- Fix: Fixed input data files and tables
- Sorc: Source code for models and utilities
- Modulefiles: Environment modules for different systems

Workflow Management:
- Uses Rocoto XML workflow engine
- Supports multiple HPC systems (Hera, Orion, WCOSS2, etc.)
- Runs GFS, GDAS, and GEFS forecast systems
`,
    jobs: `
Job Scripts (jobs/ directory):
- JGDAS_* : GDAS (Global Data Assimilation System) jobs
- JGFS_* : GFS (Global Forecast System) jobs  
- JGEFS_* : GEFS (Global Ensemble Forecast System) jobs
- Each job sets up environment and calls corresponding script
`,
    scripts: `
Scripts (scripts/ directory):
- exgdas_* : GDAS execution scripts
- exgfs_* : GFS execution scripts
- Implement the actual workflow logic called by jobs
`,
    configs: `
Configuration Files (parm/ directory):
- config/ : System-specific configuration files
- globus/ : Globus data transfer configurations
- product/ : Product generation parameters
- wave/ : Wave model configurations
`
  };

  const content = component && structure[component] 
    ? structure[component] 
    : structure.overview;

  return {
    content: [
      {
        type: 'text',
        text: content.trim()
      }
    ]
  };
}

/**
 * List job scripts
 */
async function listJobScripts() {
  try {
    const jobsDir = path.join(__dirname, 'jobs');
    const files = await fs.readdir(jobsDir);
    const jobScripts = files
      .filter(file => file.startsWith('J') && !file.includes('.'))
      .sort();

    const jobList = jobScripts.map(job => `- ${job}`).join('\n');
    
    return {
      content: [
        {
          type: 'text',
          text: `Available Job Scripts:\n\n${jobList}`
        }
      ]
    };
  } catch (error) {
    return {
      content: [
        {
          type: 'text',
          text: `Error listing job scripts: ${error.message}`
        }
      ]
    };
  }
}

/**
 * Get system configuration information
 */
async function getSystemConfigs(system) {
  const systemInfo = {
    hera: 'NOAA RDHPCS Hera - Primary development and testing system',
    orion: 'NOAA RDHPCS Orion - Research and development system',
    hercules: 'NOAA RDHPCS Hercules - Newest HPC system',
    wcoss2: 'NOAA WCOSS2 - Operational production system (Dogwood/Cactus)',
    gaeac5: 'NOAA Cloud - AWS-based compute environment',
    gaeac6: 'NOAA Cloud - Azure-based compute environment'
  };

  const info = system && systemInfo[system] 
    ? systemInfo[system]
    : 'Available systems: ' + Object.keys(systemInfo).join(', ');

  const configPath = system ? `env/${system.toUpperCase()}.env` : 'env/';
  
  return {
    content: [
      {
        type: 'text',
        text: `System: ${system || 'All systems'}\n\n${info}\n\nConfiguration files: ${configPath}`
      }
    ]
  };
}

/**
 * Explain workflow components
 */
async function explainWorkflowComponent(component) {
  const explanations = {
    rocoto: 'Rocoto is the workflow engine that manages job dependencies and execution scheduling',
    gsi: 'Gridpoint Statistical Interpolation - the data assimilation system',
    ufs: 'Unified Forecast System - the atmospheric model component',
    mom6: 'Modular Ocean Model version 6 - ocean model component',
    cice: 'Community Ice CodE - sea ice model component',
    ww3: 'WaveWatch III - ocean wave model component',
    gefs: 'Global Ensemble Forecast System - ensemble prediction system',
    gdas: 'Global Data Assimilation System - analysis and reanalysis system',
    gfs: 'Global Forecast System - deterministic forecast system'
  };

  const explanation = explanations[component?.toLowerCase()] || 
    `Component "${component}" not found. Available components: ${Object.keys(explanations).join(', ')}`;

  return {
    content: [
      {
        type: 'text', 
        text: explanation
      }
    ]
  };
}

/**
 * Main function to start the server
 */
async function main() {
  const server = createServer();
  const transport = new StdioServerTransport();
  await server.connect(transport);
  console.error('Global Workflow MCP Server running on stdio');
}

main().catch((error) => {
  console.error('Server error:', error);
  process.exit(1);
});
